### PR TITLE
Automate the since updater

### PR DESCRIPTION
**Pull request by NotMyFault**
The change proposed aims to automate the process of filling in since annotations to make PRs like https://github.com/jenkinsci/jenkins/pull/7239 redundant.

### Testing done

A PoC is available at https://github.com/timja/jenkins/pull/137

### Proposed changelog entries

- Entry 1: Issue, human-readable text
- […]

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7240"></a>

### CODE REVIEW

File: .github/workflows/run-since-updater.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+name: Run update-since-todo.py
+
+on:
+  schedule:
+    - cron: "0 16 * * WED"
```
Comment (daniel-beck) [inline .github/workflows/run-since-updater.yml:5]: Safer to do this Thursdays, since Wednesdays is when we publish weekly security updates.

Comment (NotMyFault) [reply .github/workflows/run-since-updater.yml:5]: Indeed, good catch.

Comment (NotMyFault) [inline .github/workflows/run-since-updater.yml:5]:
```suggestion
    - cron: "0 16 * * THU"
```

**Comment by daniel-beck** [comment]
The linked PR has messed up formatting, it should be a nested list. Additionally, the first output (before "List of commits") probably shouldn't be in the PR.

**Comment by NotMyFault** [comment]
> The linked PR has messed up formatting, it should be a nested list

According to the GH forums, the best you can do to preserve a bit of formatting is
https://github.com/jenkinsci/jenkins/pull/7240/files#diff-a5957abd073e9c3e3e614f9bf230c7779a16f3daba698c896f6756b7cb7d8f3eR21-R23

:/ 

> probably shouldn't be in the PR

Which part do you mean in particular?

**Comment by lemeurherve** [comment]
> According to the GH forums

Got a link?

**Comment by lemeurherve** [comment]
Note: updatecli can run scripts too, maybe another way to preserve formatting.

**Comment by NotMyFault** [comment]
> > According to the GH forums
> 
> Got a link?

https://github.com/peter-evans/create-pull-request/blob/main/docs/examples.md#setting-the-pull-request-body-from-a-file

**Comment by daniel-beck** [comment]
> Which part do you mean in particular?

Everything above the quoted string is debug output generated immediately. If the script aborts, it helps determining where it failed. Everything starting from the quoted string is output intended for convenient PR use.

**Review by timja — APPROVED** [review]
worth a go, looks nice!

File: .github/workflows/run-since-updater.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    if: ${{ github.repository_owner == 'jenkinsci' }}
+    steps:
+      - uses: actions/checkout@v3
+      - name: Run update-since-todo.py
+        run: |
+          body=$(./update-since-todo.py)
+          body="${body//'%'/'%25'}"
+          body="${body//$'\n'/'%0A'}"
+          body="${body//$'\r'/'%0D'}" 
+          echo ::set-output name=PROGRESS::$body
```
Comment (NotMyFault) [inline .github/workflows/run-since-updater.yml:24]:
```suggestion
          echo "PROGRESS=$body" >> $GITHUB_OUTPUT
```

**Comment by timja** [comment]
Multiline strings are now documented at https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#example-of-a-multiline-string

Not super clean but it works and is nicer than before.

But GitHub PR descriptions don't handle tab characters so I've turned it into a markdown list instead

@daniel-beck any thoughts now? see https://github.com/timja/jenkins/pull/137

**Comment by daniel-beck** [comment]
@timja "List of commits" looks good now 👍  Could you wrap the "Analyzing" bits in a `<details>` block as suggested in https://github.com/jenkinsci/jenkins/pull/7240#issuecomment-1272975667? Since this is just debug output I doubt we need to always see it.

**Comment by timja** [comment]
> @timja "List of commits" looks good now 👍 Could you wrap the "Analyzing" bits in a `<details>` block as suggested in [#7240 (comment)](https://github.com/jenkinsci/jenkins/pull/7240#issuecomment-1272975667)? Since this is just debug output I doubt we need to always see it.

If I understood you right that's been done, see https://github.com/timja/jenkins/pull/137

**Review by daniel-beck — APPROVED** [review]

**Pull request merged**